Content.Server/GameObjects/EntitySystems/GhostRoleSystem.cs

```diff
@@ -0,0 +1,144 @@
+using System.Collections.Generic;
+using System.Drawing;
+using Content.Server.Administration;
+using Content.Server.Eui;
+using Content.Server.GameObjects.Components.Observer;
+using Content.Shared.GameObjects.Components.Observer;
+using Content.Shared.GameObjects.EntitySystemMessages;
+using Content.Shared.GameTicking;
+using JetBrains.Annotations;
+using Robust.Server.GameObjects;
+using Robust.Server.Interfaces.Player;
+using Robust.Shared.Console;
+using Robust.Shared.GameObjects.Systems;
+using Robust.Shared.IoC;
+using Robust.Shared.Utility;
+using Robust.Shared.ViewVariables;
+
+namespace Content.Server.GameObjects.EntitySystems
+{
+    [UsedImplicitly]
+    public class GhostRoleSystem : EntitySystem, IResettingEntitySystem
+    {
+        [Dependency] private readonly EuiManager _euiManager = default!;
+
+        private uint _nextRoleIdentifier = 0;
+        private readonly Dictionary<uint, GhostRoleComponent> _ghostRoles = new();
+        private readonly Dictionary<IPlayerSession, GhostRolesEui> _openUis = new();
+
+        [ViewVariables]
+        public IReadOnlyCollection<GhostRoleComponent> GhostRoles => _ghostRoles.Values;
+
+        public override void Initialize()
+        {
+            SubscribeLocalEvent<PlayerAttachSystemMessage>(OnPlayerAttached);
+        }
+
+        private uint GetNextRoleIdentifier()
+        {
+            return unchecked(_nextRoleIdentifier++);
+        }
+
+        public void OpenEui(IPlayerSession session)
+        {
+            if (session.AttachedEntity == null || !session.AttachedEntity.HasComponent<GhostComponent>())
+                return;
+
+            if(_openUis.ContainsKey(session))
+                CloseEui(session);
+
+            var eui = _openUis[session] = new GhostRolesEui();
+            _euiManager.OpenEui(eui, session);
+            eui.StateDirty();
```

Comment:
If a new Eui is made every time it's opened should the events in it be unsubscribed? So the ref gets GC'd.

Reply:
As discussed on discord, this should be fine as all references to the EUI are removed when it's closed.

Reply:
As discussed on discord, this should be fine as all references to the EUI are removed when it's closed.

Reply:
As discussed on discord, this should be fine as all references to the EUI are removed when it's closed.